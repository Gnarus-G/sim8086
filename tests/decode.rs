use std::{io::Write, process::Command};

use insta::assert_display_snapshot;

fn test_with(file: &str) {
    let _ = Command::new("nasm")
        .arg(format!("./{}.asm", file))
        .status()
        .unwrap();

    let app_output = assert_cmd::Command::cargo_bin("sim8086")
        .unwrap()
        .arg(format!("./{}", file))
        .output()
        .map(|out| String::from_utf8(out.stdout).unwrap())
        .unwrap();

    let test_file_path = format!("./{}_test.asm", file);

    let mut test_file = std::fs::File::create(&test_file_path).unwrap();

    write!(test_file, "{}", app_output).unwrap();

    let _ = Command::new("nasm").arg(test_file_path).status().unwrap();

    assert_cmd::Command::new("diff")
        .arg(format!("./{}", file))
        .arg(format!("./{}_test", file))
        .assert()
        .success();

    assert_display_snapshot!(app_output);
}

#[test]
fn movs() {
    test_with("many_register_mov");

    test_with("more_movs");

    test_with("challenge_movs");
}

#[test]
fn add_sub_cmp() {
    test_with("add_sub_cmp");
}
