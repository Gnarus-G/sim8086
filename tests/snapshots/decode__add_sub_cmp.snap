---
source: tests/decode.rs
description: add_sub_cmp
expression: app_output
---
bits 16

add bx, [bx + si]
add bx, [bp + 0]
add si, word 2
add bp, word 2
add cx, word 8
add bx, [bp + 0]
add cx, [bx + 2]
add bh, [bp + si + 4]
add di, [bp + di + 6]
add [bx + si], bx
add [bp + 0], bx
add [bp + 0], bx
add [bx + 2], cx
add [bp + si + 4], bh
add [bp + di + 6], di
add [bx], byte 34
add [bp + si + 1000], word 29
add ax, [bp + 0]
add al, [bx + si]
add ax, bx
add al, ah
add ax, 1000
add al, 226
add al, 9
sub bx, [bx + si]
sub bx, [bp + 0]
sub si, word 2
sub bp, word 2
sub cx, word 8
sub bx, [bp + 0]
sub cx, [bx + 2]
sub bh, [bp + si + 4]
sub di, [bp + di + 6]
sub [bx + si], bx
sub [bp + 0], bx
sub [bp + 0], bx
sub [bx + 2], cx
sub [bp + si + 4], bh
sub [bp + di + 6], di
sub [bx], byte 34
sub [bx + di], word 29
sub ax, [bp + 0]
sub al, [bx + si]
sub ax, bx
sub al, ah
sub ax, 1000
sub al, 226
sub al, 9

